src/btree/bt_read.c

```diff
@@ -279,22 +278,20 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
               F_ISSET(session, WT_SESSION_READ_WONT_NEED) ||
               F_ISSET(S2C(session)->cache, WT_CACHE_EVICT_NOKEEP);
             continue;
-        case WT_REF_READING:
-            if (LF_ISSET(WT_READ_CACHE))
-                return (WT_NOTFOUND);
-            if (LF_ISSET(WT_READ_NO_WAIT))
-                return (WT_NOTFOUND);
-
-            /* Waiting on another thread's read, stall. */
-            WT_STAT_CONN_INCR(session, page_read_blocked);
-            stalled = true;
-            break;
         case WT_REF_LOCKED:
             if (LF_ISSET(WT_READ_NO_WAIT))
                 return (WT_NOTFOUND);
 
-            /* Waiting on eviction, stall. */
-            WT_STAT_CONN_INCR(session, page_locked_blocked);
+            if (F_ISSET(ref, WT_REF_READING)) {
```

Comment:
Can the ref have gone from WT_REF_LOCKED to WT_REF_MEM with a valid page modify structure between those two checks? Would it be OK to skip the page in that case?

I think it probably is for checkpoint - that change must happen after the checkpoint started, so shouldn't need to be included. I'm less sure about __wt_evict_file, but I suspect that's OK too.

Reply:
FWIW, I think the check here is safe because `WT_REF_READING` is only set while the ref stays in the `WT_REF_LOCKED` state.  So seeing it set here implies that the page is being read and thus is safe to skip (when `WT_READ_CACHE` is set).

Reply:
> Can the ref have gone from WT_REF_LOCKED to WT_REF_MEM with a valid page modify structure between those two checks? Would it be OK to skip the page in that case?

Yes.

Basically, there's a race here (and there was a race before), it's just a different set of races, that is, we could race between seeing `WT_REF_READING` and stalling based on the page being read, or, in this case, between seeing `WT_REF_LOCKED` and checking the new "is-reading" flag.

I would agree I've made the race worse in this case (I'm checking 2 fields, rather than the 1 field checked before), but I don't think it matters. In all cases, we're going to stall when it might not have been necessary.